### PR TITLE
*: scatter the leader distribution in the specified range

### PR DESCRIPTION
**Pull request by nolouch**
add a scheduler for balance leader distribution in a range.

### CODE REVIEW

File: server/schedulers/scatter_range_leader.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+				loopEnd = true
+				break
+			}
+			if string(r.EndKey) == "" {
+				loopEnd = true
+				break
+			}
+			startKey = r.EndKey
+		}
+
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: please delete redundant line

File: server/schedulers/scatter_range_leader.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	peer, err := cluster.AllocPeer(target.GetId())
+	if err != nil {
+		if _, ok := followers[target.GetId()]; ok {
+			step := schedule.TransferLeader{FromStore: source.GetId(), ToStore: target.GetId()}
+			op := schedule.NewOperator("scatter-range-leader", sourceRegion.GetId(), schedule.OpRange|schedule.OpLeader, step)
+			return []*schedule.Operator{op}
+		}
+		return nil
+	}
+	steps = append(steps, schedule.AddPeer{ToStore: target.GetId(), PeerID: peer.GetId()})
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: I prefer appending all steps in one append().

File: server/schedulers/scatter_range_leader.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	maxScore := int64(math.MinInt32)
+	minScore := int64(math.MaxInt32)
+	for _, s := range stores {
+		score := regions.GetStoreLeaderRegionSize(s.GetId())
+		if score > maxScore {
+			maxScore = score
+			source = s
+		}
+	}
+	for _, s := range stores {
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: please extract these logics to a new selector.

File: server/schedulers/scatter_range_leader.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		return nil
+	}
+	steps = append(steps, schedule.AddPeer{ToStore: target.GetId(), PeerID: peer.GetId()})
+	steps = append(steps, schedule.TransferLeader{FromStore: source.GetId(), ToStore: target.GetId()})
+	steps = append(steps, schedule.RemovePeer{FromStore: source.GetId()})
+	op := schedule.NewOperator("scatter-range-leader", sourceRegion.GetId(), schedule.OpRange|schedule.OpLeader|schedule.OpBalance, steps...)
+	return []*schedule.Operator{op}
+}
+func (l *scatterRangeLeaderScheduler) shouldBalance(sourceID, targetID uint64, sourceRegion *core.RegionInfo, regions *core.RegionsInfo) bool {
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: please add blank line between functions.

File: server/schedulers/scatter_range_leader.go (hunk 2 above)
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: I prefer appending all steps in one append().

File: server/schedulers/scatter_range_leader.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+	var steps []schedule.OperatorStep
+	followers := sourceRegion.GetFollowers()
+	for storeID := range followers {
+		if l.shouldBalance(source.GetId(), storeID, sourceRegion, regions) {
+			target = cluster.GetStore(storeID)
+			break
+		}
+	}
+	if _, ok := followers[target.GetId()]; ok {
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: these can be removed to if body of l.shouldBalance.

File: server/schedulers/scatter_range_leader.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		step := schedule.TransferLeader{FromStore: source.GetId(), ToStore: target.GetId()}
+		op := schedule.NewOperator("scatter-range-leader", sourceRegion.GetId(), schedule.OpRange|schedule.OpLeader, step)
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: add metrics.

File: server/schedulers/scatter_range_leader.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		return []*schedule.Operator{op}
+	}
+	peer, err := cluster.AllocPeer(target.GetId())
+	if err != nil {
+		if _, ok := followers[target.GetId()]; ok {
```
Comment (Connor1996) [inline server/schedulers/scatter_range_leader.go]: if target is in followers, it can't reach here.

File: server/schedulers/scatter_range.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+type scatterRangeScheduler struct {
+	*baseScheduler
+	rangeName     string
+	filters       []schedule.Filter
+	startKey      []byte
+	endKey        []byte
+	balanceLeader schedule.Scheduler
+	balanceRegion schedule.Scheduler
+	operators     []*schedule.Operator
```
Comment (Connor1996) [inline server/schedulers/scatter_range.go]: no use?

File: server/schedulers/scatter_range.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+func (l *scatterRangeScheduler) Schedule(cluster schedule.Cluster, opInfluence schedule.OpInfluence) []*schedule.Operator {
+	schedulerCounter.WithLabelValues(l.GetName(), "schedule").Inc()
+	c := schedule.GenRangeCluster(cluster, l.startKey, l.endKey)
+	c.SetTolerantSizeRatio(2)
+	influence := l.getOperators(opInfluence)
+	ops := l.balanceLeader.Schedule(c, schedule.NewOpInfluence(influence, cluster))
+	if len(ops) > 0 {
+		ops[0].SetDesc(fmt.Sprintf("scatter-range-leader-%s", l.rangeName))
+		ops[0].AttachKind(schedule.OpRange)
```
Comment (Connor1996) [inline server/schedulers/scatter_range.go]: please add metrics

File: server/schedulers/scatter_range.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+		return ops
+	}
+	ops = l.balanceRegion.Schedule(c, schedule.NewOpInfluence(influence, cluster))
+	if len(ops) > 0 {
+		ops[0].SetDesc(fmt.Sprintf("scatter-range-region-%s", l.rangeName))
+		ops[0].AttachKind(schedule.OpRange)
```
Comment (Connor1996) [inline server/schedulers/scatter_range.go]: ditto

File: server/schedulers/scatter_range.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+		return ops
+	}
+	return nil
```
Comment (Connor1996) [inline server/schedulers/scatter_range.go]: ditto

File: server/schedule/operator_kind.go [hunk 12]
```diff
@@ -32,6 +32,7 @@ const (
 	OpReplica                            // Initiated by replica checkers.
 	OpBalance                            // Initiated by balancers.
 	OpMerge                              // Initiated by merge checkers.
+	OpRange
```
Comment (Connor1996) [inline server/schedule/operator_kind.go]: add comment

File: server/schedule/operator.go [hunk 13]
```diff
@@ -275,6 +275,16 @@ func (o *Operator) Desc() string {
 	return o.desc
 }
 
+// SetDesc set the description for the operator.
```
Comment (Connor1996) [inline server/schedule/operator.go]: s/set/sets

File: server/schedule/operator.go [hunk 14]
```diff
@@ -275,6 +275,16 @@ func (o *Operator) Desc() string {
...
+func (o *Operator) SetDesc(desc string) {
+	o.desc = desc
+}
+
+// AttachKind attach a operator kind for the operator.
```
Comment (Connor1996) [inline server/schedule/operator.go]: s/attach a/attaches an

**Comment by disksing** [comment]
LGTM.

**Pull request merged**